osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/ReplaceKeyCommandExecutor.java

```diff
@@ -130,17 +130,30 @@ public DlmsDevice execute(
       final MessageMetadata messageMetadata)
       throws OsgpException {
 
-    if (!keyWrapper.isGenerated()) {
+    ReplaceKeyInput keyWrapper2;
+    if (keyWrapper.isGenerated()) {
+      keyWrapper2 = keyWrapper;
+    } else {
       this.secretManagementService.storeNewKey(
           messageMetadata,
           device.getDeviceIdentification(),
           keyWrapper.getSecurityKeyType(),
           keyWrapper.getBytes());
+
+      final byte[] newKey =
+          this.secretManagementService.getNewKey(
+              messageMetadata, device.getDeviceIdentification(), keyWrapper.getSecurityKeyType());
+      keyWrapper2 =
+          ReplaceKeyInput.from(
+              newKey,
+              keyWrapper.getKeyId(),
+              keyWrapper.getSecurityKeyType(),
+              keyWrapper.isGenerated());
```

Comment:
Why is this needed and why is it done for a key from a replace key request and not for one that was generated?
As I understand it, the `keyWrapper` always has the plaintext (not in any way encrypted) bytes for the key.
Perhaps there is a good reason but it is not immediately obvious to me looking at this code.
It might be related to the logic in store keys that favors older keys over more recent ones, but I have a suspicion that using the newest value is more logical and leads to simpler code (the extra call here to get the new key would not be needed).
I believe either other refactorings or adding some comments explaining the logic here could go a long way in helping someone who will need to work in this area in the future.
It would probably also be nice to mention that `secretManagementService.storeNewKey` is not called for a generated key, because generated keys are stored with the secret management service upon generation, referring to `GenerateAndReplaceKeyCommandExecutor`.

Reply:
I tend to agree with you  but s said before the story _explicitely states that the NEW key in the database (most recent if multiple) is used to be transferred to the device._
I agree that some documentation might need to explain this code

Reply:
REvision of requirements in story: NEW key sent with the request is always used. Existing NEW keys are set to status WITHDRAWN